src/Paprika/Data/SlottedArray.cs

```diff
@@ -904,80 +908,199 @@ private void MarkAsDeleted(int index)
+                    deletedMask ^= 1U << setBitIndex;
+                    alive--;
+                    startIndex++;
+                    endIndex = setBitIndex + (i - HashesPerVector) / VectorsByBatch - 1;
                 }
 
-                // Copy hash
-                GetHashRef(writeAt) = GetHashRef(i);
+                if (startIndex == endIndex + 1)
+                {
+                    Debug.Assert(deletedMask == 0, "Did not consume all the set bits");
+
+                    // Could not find any non-consecutive set bits, progress start to the next non deleted slot.
+                    startIndex++;
+
+                    // Move on to the next vector to find the new end.
+                    break;
+                }
 
-                // Copy everything, just overwrite the address
-                ref var destinationSlot = ref GetSlotRef(writeAt);
-                destinationSlot.KeyPreamble = slot.KeyPreamble;
-                destinationSlot.ItemAddress = writtenTo;
+                writeAtIndex = CopyDataInternal(startIndex, endIndex, writeAtIndex);
+                startIndex = endIndex + 2;
+            }
+        }
 
-                writeAt++;
+        // If there was no valid end found for a corresponding start, move all the remaining slot.
+        if (startIndex != NotFound && startIndex < count)
+        {
+            while (GetSlotRef(startIndex).IsDeleted && startIndex < count)
+            {
+                startIndex++;
             }
 
-            // Memoize to what is read to
-            readTo = addr;
+            endIndex = count - 1;
+
+            if (startIndex <= endIndex)
+            {
+                writeAtIndex = CopyDataInternal(startIndex, endIndex, writeAtIndex);
+            }
         }
 
-        // Finalize by setting the header
+        Debug.Assert(_header.Deleted == Count - alive, "All the deleted items were not discovered");
+        Debug.Assert(writeAtIndex != NotFound, "No deleted slot was found");
+
+        // Adjust header values
         _header.Low = (ushort)(alive * Slot.TotalSize);
-        _header.High = (ushort)(_data.Length - writtenTo);
         _header.Deleted = 0;
+        _header.High = (ushort)((writeAtIndex == 0)
+            ? 0
+            : _data.Length - GetSlotRef(writeAtIndex - 1).ItemAddress);
+    }
+
+    /// <summary>
+    /// Helper function to create a deleted bitmask from the underlying data where each bit is
+    /// either 1 (deleted) or 0 (alive) for the corresponding slot index.
+    /// </summary>
+    private uint CreateDeletedBitmask(ref ushort d, int offset)
+    {
+        uint deletedMask = 0;
+
+        // Extract the preamble data from the underlying slots and compare it with the preamble delete mask.
+        if (Vector256.IsHardwareAccelerated)
+        {
+            var slotData = Vector256.LoadUnsafe(ref d, (UIntPtr)offset);
+            var preambleData = Vector256.BitwiseAnd(slotData, Slot.GetKeyPreambleMaskAsVector256());
```

Comment:
This loads the data, then bitwise and it with the mask, then compares. Should we have it first checked if it's non-zero? I believe this is the approach taken by the `Span.IndexOf` as getting the significant bits where the value is zero is a bit more expensive. Again, to compare with the benchmarks